File: buggy-api/question.py
from collections import defaultdict
import pprint # Using pretty-print for cleaner output

# --- GIVEN CODE (Mock API Service) ---
# (You do not need to modify this part.)

def get_transactions_from_api() -> list[dict]:
    """
    Simulates fetching data from a buggy external API.
    """
    print("Fetching data from buggy API...")
    return [
        {"user_id": 101, "type": "credit", "amount": 50.00},
        {"user_id": 102, "type": "debit",  "amount": 75.20},
        {"user_id": 101, "type": "debit",  "amount": "120.50"}, # BUG 1: Amount is a string
        {"user_id": 103, "type": "credit", "amount": 200.00},
        {"user_id": 101, "type": "debit",  "amount": 30.00},
        {"user_id": 102, "type": "credit", "amount": None},      # BUG 2: Amount is None
        {"user_id": 101},                                      # BUG 3: Missing 'amount' and 'type'
        {"user_id": 103, "type": "refun",  "amount": 10.00},     # BUG 4: Invalid 'type'
        {"user_id": 102, "type": "debit",  "amount": -25.00},    # BUG 5: Invalid (negative) 'amount'
    ]

# --- EXAMINEE'S TASK ---

def process_transactions(transactions: list[dict]) -> tuple[dict, list[dict]]:
    """
    Implement this function to process the raw transaction data.

    The function must return a tuple containing two items:
    1.  `summary_report`: A dictionary (same as before) with user_ids as keys
        and their 'total_debit' / 'total_credit' as values.
    2.  `error_report`: A list of dictionaries. Each dictionary should
        log an error, capturing the original index of the transaction
        and a clear message about what was wrong.
        e.g., {'index': 6, 'message': 'Missing required key: amount'}

    Business Rules for Validation:
    - A transaction is INVALID if:
        1. It is missing 'user_id', 'type', or 'amount'.
        2. 'type' is not *exactly* "credit" or "debit".
        3. 'amount' cannot be converted to a positive float.
    - If a transaction is INVALID, it should be skipped in the summary
      and an entry should be added to the error_report.
    - If a transaction is VALID, it should be added to the summary_report.
    """
    
    # Using defaultdict for cleaner summary logic
    summary_report = defaultdict(lambda: {"total_debit": 0.0, "total_credit": 0.0})
    error_report = []
    
    # TODO: Implement your logic here.
    # You will need to iterate with an index (e.g., using enumerate)
    
    
    # Convert defaultdict back to a regular dict for the final output
    return dict(summary_report), error_report

# --- Main execution for testing ---
raw_data = get_transactions_from_api()
summary, errors = process_transactions(raw_data)

print("\n--- Processed Summary ---")
pprint.pprint(summary)
print("\n--- Error Report ---")
pprint.pprint(errors)

File: buggy-api/question.cs
using System;
using System.Collections.Generic;

// --- GIVEN CODE (Mock API Service & Data Models) ---
// (You do not need to modify this part.)

public class ApiTransaction
{
    public int? UserId { get; set; } // Now nullable
    public string Type { get; set; }
    public object Amount { get; set; }
}

public class TransactionService
{
    public List<ApiTransaction> GetTransactionsFromApi()
    {
        Console.WriteLine("Fetching data from buggy API...");
        return new List<ApiTransaction>
        {
            new ApiTransaction { UserId = 101, Type = "credit", Amount = 50.00d },
            new ApiTransaction { UserId = 102, Type = "debit",  Amount = 75.20d },
            new ApiTransaction { UserId = 101, Type = "debit",  Amount = "120.50" }, // BUG 1: String
            new ApiTransaction { UserId = 103, Type = "credit", Amount = 200.00d },
            new ApiTransaction { UserId = 101, Type = "debit",  Amount = 30.00d },
            new ApiTransaction { UserId = 102, Type = "credit", Amount = null },      // BUG 2: Null Amount
            new ApiTransaction { UserId = 101, Type = null,    Amount = 20.00d },  // BUG 3: Null Type
            new ApiTransaction { UserId = 103, Type = "refun",  Amount = 10.00d },    // BUG 4: Invalid Type
            new ApiTransaction { UserId = 102, Type = "debit",  Amount = -25.00d },   // BUG 5: Negative Amount
            new ApiTransaction { UserId = null, Type = "credit",Amount = 100.00d }, // BUG 6: Null UserId
        };
    }
}

// Clean summary data structure
public class UserSummary
{
    public double TotalDebit { get; set; } = 0.0;
    public double TotalCredit { get; set; } = 0.0;
}

// New model for reporting an error
public class ProcessingError
{
    public int OriginalRecordIndex { get; set; }
    public string ErrorMessage { get; set; }
}

// New model for the final result
public class IngestionResult
{
    public Dictionary<int, UserSummary> SummaryReport { get; set; }
    public List<ProcessingError> ErrorReport { get; set; }

    public IngestionResult()
    {
        SummaryReport = new Dictionary<int, UserSummary>();
        ErrorReport = new List<ProcessingError>();
    }
}

// --- EXAMINEE'S TASK ---

public class TransactionProcessor
{
    public IngestionResult ProcessTransactions(List<ApiTransaction> transactions)
    {
        /*
        Implement this function to process the raw transaction data.

        The function must return an 'IngestionResult' object which contains:
        1.  'SummaryReport': A Dictionary with user_ids (int) as keys
            and 'UserSummary' objects as values.
        2.  'ErrorReport': A List of 'ProcessingError' objects.

        Business Rules for Validation:
        - A transaction is INVALID if:
            1. 'UserId', 'Type', or 'Amount' is null.
            2. 'Type' is not *exactly* "credit" or "debit".
            3. 'Amount' cannot be converted to a positive double.
        - If a transaction is INVALID, it should be skipped in the summary
          and a new 'ProcessingError' should be added to the ErrorReport.
        - If a transaction is VALID, it should be added to the SummaryReport.
        */
        
        var result = new IngestionResult();
        
        // TODO: Implement your logic here.
        // You will need a 'for' loop to get the index.
        

        return result;
    }
}

// --- Main execution for testing ---
// (In a real test, you'd have this in a separate Main method)
/*
var service = new TransactionService();
var processor = new TransactionProcessor();
var rawData = service.GetTransactionsFromApi();
var report = processor.ProcessTransactions(rawData);

Console.WriteLine("\n--- Processed Summary ---");
foreach (var entry in report.SummaryReport)
{
    Console.WriteLine($"User {entry.Key}: Debit={entry.Value.TotalDebit}, Credit={entry.Value.TotalCredit}");
}

Console.WriteLine("\n--- Error Report ---");
foreach (var error in report.ErrorReport)
{
    Console.WriteLine($"Index {error.OriginalRecordIndex}: {error.ErrorMessage}");
}
*/
